File: creational/singleton/configuration_manager/python/configuration_manager.py
import threading

class ConfigurationManager:
    """
    Singleton Configuration Manager using a lock for thread safety.

    This implementation uses a class variable to hold the single instance
    and a lock to ensure thread-safe initialization.
    """
    _instance = None
    _lock = threading.Lock()
    _config_data = {}

    def __new__(cls):
        # Double-checked locking pattern
        if cls._instance is None:
            with cls._lock:
                # Check again inside the lock
                if cls._instance is None:
                    cls._instance = super(ConfigurationManager, cls).__new__(cls)
                    # Initialize configuration (e.g., load from file/env)
                    cls._instance._load_config()
        return cls._instance

    def _load_config(self):
        """Simulates loading configuration data."""
        print("Loading configuration...")
        # In a real app, load from a file, database, or environment variables
        self._config_data = {
            "api_key": "dummy_api_key_123",
            "database_url": "localhost:5432",
            "feature_flag_x": True
        }
        print("Configuration loaded.")

    def get_setting(self, key):
        """Get a configuration setting by key."""
        return self._config_data.get(key)

    def set_setting(self, key, value):
        """Set a configuration setting."""
        self._config_data[key] = value

    def get_all_settings(self):
        """Return all configuration settings."""
        return self._config_data.copy()

File: creational/singleton/configuration_manager/typescript/configuration_manager.ts
/**
 * Singleton Configuration Manager using a private static instance.
 *
 * This implementation relies on TypeScript's class features and a private
 * static member to hold the single instance.
 */
export class ConfigurationManager {
  private static instance: ConfigurationManager | null = null;
  private configData: { [key: string]: any } = {};
  private isLoaded: boolean = false;

  // Private constructor ensures no external instantiation
  private constructor() {
    this.loadConfig();
  }

  /**
   * Gets the single instance of the ConfigurationManager.
   * Creates the instance if it doesn't exist yet.
   */
  public static getInstance(): ConfigurationManager {
    if (ConfigurationManager.instance === null) {
      ConfigurationManager.instance = new ConfigurationManager();
    }
    return ConfigurationManager.instance;
  }

  /**
   * Simulates loading configuration data.
   * In a real app, this could read from a file, env vars, or a config service.
   */
  private loadConfig(): void {
    if (!this.isLoaded) {
      console.log("Loading configuration...");
      // Simulate loading data
      this.configData = {
        apiKey: "dummy_api_key_ts_456",
        serverUrl: "https://api.example-ts.com",
        featureFlagY: false,
      };
      this.isLoaded = true;
      console.log("Configuration loaded.");
    }
  }

  /**
   * Gets a configuration setting by key.
   * @param key - The configuration key.
   * @returns The value associated with the key, or undefined if not found.
   */
  public getSetting(key: string): any {
    return this.configData[key];
  }

  /**
   * Sets a configuration setting.
   * @param key - The configuration key.
   * @param value - The value to set.
   */
  public setSetting(key: string, value: any): void {
    this.configData[key] = value;
  }

  /**
   * Returns a copy of all configuration settings.
   */
  public getAllSettings(): { [key: string]: any } {
    // Return a copy to prevent external modification of the internal state
    return { ...this.configData };
  }
}
